# SQL

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Цели исследования
- проанализировать базу данных
- решить поставленные задачи
- изучить текущую ситуацию на книжном рынке для создания нового продукта

## Доступ к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})


**Обзор таблиц**

In [2]:
# смотрим первые 5 записей в таблице books
books  = """
SELECT *
FROM books
LIMIT 5
"""
con=engine.connect()

pd.io.sql.read_sql(sql=text(books), con = con)


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [3]:
# смотрим первые 5 записей в таблице authors
authors  = """
SELECT *
FROM authors
LIMIT 5
"""
pd.io.sql.read_sql(sql=text(authors), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [4]:
# смотрим первые 5 записей в таблице publishers
publishers  = """
SELECT *
FROM publishers
LIMIT 5
"""
pd.io.sql.read_sql(sql=text(publishers), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [5]:
# смотрим первые 5 записей в таблице ratings
ratings  = """
SELECT *
FROM ratings
LIMIT 5
"""
pd.io.sql.read_sql(sql=text(ratings), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [6]:
# смотрим первые 5 записей в таблице reviews
reviews  = """
SELECT *
FROM reviews
LIMIT 5
"""
pd.io.sql.read_sql(sql=text(reviews), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

# Задания

**Посчитайте, сколько книг вышло после 1 января 2000 года**

In [7]:
query_1  = """
SELECT COUNT(book_id)
FROM books
WHERE publication_date >= '2000-01-01';
"""
pd.io.sql.read_sql(sql=text(query_1), con = con)

count
0    821

С 1 января 2000 года была выпущена 821 книга.

**Для каждой книги посчитайте количество обзоров и среднюю оценку**

In [8]:
query_2  = """
SELECT b.book_id,
b.title,
COUNT(DISTINCT r.review_id),
ROUND(AVG(rat.rating),2) AS avg_rating
FROM books AS b
LEFT JOIN reviews AS r ON b.book_id = r.book_id
LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
GROUP BY b.book_id, b.title
ORDER BY count DESC, avg_rating DESC;
"""
pd.io.sql.read_sql(sql=text(query_2), con = con)

book_id                                              title  count  \
0        948                            Twilight (Twilight  #1)      7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...      6   
2        299  Harry Potter and the Chamber of Secrets (Harry...      6   
3        656                                     The Book Thief      6   
4        734                                   The Glass Castle      6   
..       ...                                                ...    ...   
995      191  Disney's Beauty and the Beast (A Little Golden...      0   
996      387                               Leonardo's Notebooks      0   
997      221                          Essential Tales and Poems      0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel      0   
999      808                            The Natural Way to Draw      0   

     avg_rating  
0          3.66  
1          4.41  
2          4.29  
3          4.26  
4          4.21  
..          ...  
995        4.00  
996        4.00  
997        4.00  
998        3.67  
999        3.00  

[1000 rows x 4 columns]

Больше всего обзоров на книгу `Twilight`, ее средний рейтинг составляет 3.66, далее следуют `Harry Potter and the Prisoner of Azkaban` и `Harry Potter and the Chamber of Secrets` с рейтингом 4.41 и 4.29 соответственно. Также в ТОП-5 книг по обзорам вошли `The Book Thief` и `The Glass Castle` с рейтингом 4.26 и 4.21 соответственно.

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры**

In [9]:
query_3  = """
SELECT p.publisher,
COUNT(b.book_id)
FROM books AS b
LEFT JOIN publishers AS p ON b.publisher_id = p.publisher_id
WHERE b.num_pages > 50
GROUP BY p.publisher
ORDER BY count DESC
LIMIT 1;
"""
pd.io.sql.read_sql(sql=text(query_3), con = con)

publisher  count
0  Penguin Books     42

Издательство `Penguin Books` выпустило наибольшее число книг толще 50 страниц (42 книги).

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками**

In [10]:
query_4  = """

SELECT a.author,
ROUND(AVG(rat.rating),2) AS avg_rating
FROM books AS b
LEFT JOIN authors AS a ON b.author_id = a.author_id
LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
WHERE b.book_id IN (
    SELECT b.book_id
    FROM books AS b
    LEFT JOIN ratings AS rat ON b.book_id = rat.book_id
    GROUP BY b.book_id
    HAVING COUNT(rating_id) >= 50
    ORDER BY COUNT(rating_id) DESC)
GROUP BY a.author
ORDER BY avg_rating DESC
LIMIT 1
;
"""
pd.io.sql.read_sql(sql=text(query_4), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré        4.29

Автором с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками, является `J.K. Rowling` - средняя оценка книг 4.29

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок**

In [11]:
query_5  = """
SELECT AVG(count) AS avg_review
FROM (
    SELECT r.username,
    COUNT(DISTINCT r.review_id)
    FROM reviews AS r
    FULL OUTER JOIN ratings AS rat ON r.username = rat.username
    WHERE r.username IN (
        SELECT rat.username
        FROM ratings AS rat
        GROUP BY rat.username
        HAVING COUNT(rating_id) > 48
        ORDER BY COUNT(rating_id) DESC)
    GROUP BY r.username) AS c;
"""
pd.io.sql.read_sql(sql=text(query_5), con = con)

avg_review
0        24.0

Среднее кол-во обзоров от пользователей, которые поставили больше 48 оценок, равно 24.

# Вывод

Таким образом, проведя работу, мы выяснили следующее:
1. С 1 января 2000 года была выпущена 821 книга.
2. Больше всего обзоров на книгу Twilight, ее средний рейтинг составляет 3.66, далее следуют Harry Potter and the Prisoner of Azkaban и Harry Potter and the Chamber of Secrets с рейтингом 4.41 и 4.29 соответственно. Также в ТОП-5 книг по обзорам вошли The Book Thief и The Glass Castle с рейтингом 4.26 и 4.21 соответственно.
3. Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц (42 книги).
4. Автором с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками, является J.K. Rowling - средняя оценка книг 4.29
5. Среднее кол-во обзоров от пользователей, которые поставили больше 48 оценок, равно 24.